src/program.cpp

```diff
@@ -894,9 +894,92 @@ std::string cvk_program::prepare_build_options(const cvk_device* device) const {
     return options;
 }
 
+bool cvk_program::parse_user_spec_constants() {
 #if COMPILER_AVAILABLE
 #ifndef CLSPV_ONLINE_COMPILER
+    // We'll need to go through the whole temp folder rigamarole to query the
+    // spec constant info with the command line tool.
+    std::filesystem::path tmp_prefix(config.compiler_temp_dir());
+    std::filesystem::path tmp_suffix("clvk-XXXXXX");
+    std::string tmp_template = (tmp_prefix / tmp_suffix).string();
+    const char* tmp = cvk_mkdtemp(tmp_template);
+    if (tmp == nullptr) {
+        cvk_error_fn("Could not create temporary folder \"%s\"",
+                     tmp_template.c_str());
+        return false;
+    }
+    std::string tmp_folder = tmp;
+    cvk_info("Created temporary folder \"%s\"", tmp_folder.c_str());
+
+    std::string llvmspirv_input_file = tmp_folder + "/source.spv";
+
+    if (!save_il_to_file(llvmspirv_input_file, m_il)) {
+        cvk_error_fn("Couldn't save IL to file!");
+        return false;
+    }
+
+    std::string cmd_spv =
+        std::string(config.llvmspirv_bin) + " --spec-const-info";
+
+    std::string output = "";
+    cvk_exec(cmd_spv + " " + llvmspirv_input_file + " --spec-const-info",
+             &output);
+    auto output_stream = std::istringstream(output);
+
+    for (std::string line; std::getline(output_stream, line);) {
+        if (line.find("Spec const id") == std::string::npos) {
+            continue;
+        }
 
+        auto delim = line.find(",");
+        auto id_string = line.substr(0, delim);
+        line.erase(0, delim + 1);
+        id_string = id_string.substr(id_string.find("=") + 1);
+        auto id = static_cast<uint32_t>(atoi(id_string.c_str()));
+
+        delim = line.find(",");
+        auto size_string = line.substr(0, delim);
+        line.erase(0, delim + 1);
+        size_string = size_string.substr(size_string.find("=") + 1);
+        auto size = static_cast<uint32_t>(atoi(size_string.c_str()));
+
+        // We need to consume the "=" and the " " here since we aren't
+        // converting the value from a string.
+        auto type_string = line.substr(line.find("=") + 2);
+        m_user_spec_constants[id] = user_spec_constant_data(type_string, size);
```

Comment:
The lack of error handling is making me a bit nervous. Hopefully the translator does not change its output format too frequently and never produces output that would be invalid in a way that would trip this code over. Maybe a regex would be better? I think I'm fine with no change here.